controllers/utils/conditions.go

```diff
@@ -0,0 +1,152 @@
+package utils
+
+import (
+	ranv1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConditionType is a string representing the condition's type
+type ConditionType string
+
+// ConditionTypes define the different types of conditions that will be set
+var ConditionTypes = struct {
+	PrepPending         ConditionType
+	PrepSucceeded       ConditionType
+	PreUpgradePending   ConditionType
+	PreUpgradeSucceeded ConditionType
+	UpgradePending      ConditionType
+	UpgradeSucceeded    ConditionType
+	CommitPending       ConditionType
+	CommitSucceeded     ConditionType
+	RollbackPending     ConditionType
+	RollbackSucceeded   ConditionType
+}{
+	PrepPending:         "PrepPending",
+	PrepSucceeded:       "PrepSucceeded",
+	PreUpgradePending:   "PreUpgradePending",
+	PreUpgradeSucceeded: "PreUpgradeSucceeded",
+	UpgradePending:      "UpgradePending",
+	UpgradeSucceeded:    "UpgradeSucceeded",
+	CommitPending:       "CommitPending",
+	CommitSucceeded:     "CommitSucceeded",
+	RollbackPending:     "RollbackPending",
+	RollbackSucceeded:   "RollbackSucceeded",
+}
+
+// ConditionReason is a string representing the condition's reason
+type ConditionReason string
+
+// ConditionReasons define the different reasons that conditions will be set for
+var ConditionReasons = struct {
+	Completed  ConditionReason
+	Failed     ConditionReason
+	TimedOut   ConditionReason
+	InProgress ConditionReason
+	Blocked    ConditionReason
+}{
+	Completed:  "Completed",
+	Failed:     "Failed",
+	TimedOut:   "TimedOut",
+	InProgress: "InProgress",
+	Blocked:    "Blocked",
+}
+
+// SetStatusCondition is a convenience wrapper for meta.SetStatusCondition that takes in the types defined here and converts them to strings
+func SetStatusCondition(existingConditions *[]metav1.Condition, conditionType ConditionType, conditionReason ConditionReason, conditionStatus metav1.ConditionStatus, message string) {
```

Comment:
I don't think we currently have any race conditions because there is a uni-directional move through the state machine (prep -> pre -> upgrade -> rollback|commit), at least in the "sunny-day" scenario. However, do we want to set the observedGeneration field in the status on every pass through the reconcile loop (perhaps here)? It is likely easier to add now than to put it in place later.

Reply:
We can. We don't need it so much as the policy will be verifying distinct condition types. This does remind me a question. Do we allow any other spec update once the thing is started?

Reply:
Certain fields are used by different phases (eg backup crs used by pre phase). It would be user-friendly to allow these fields to be set at the same time as enabling that phase.

Reply:
Ok, we'll document this for each field then.

Reply:
`observedGeneration` updated